File: 300_longest_increasing_subsequence/Solution2/Solution.java
class Solution {
    public int lengthOfLIS(int[] nums) {
        int arrLen = nums.length;
        if (arrLen < 2) return arrLen;

        int[] tails = new int[arrLen];
        int result = 0;
        for (int num : nums) {
            int l = 0, r = result;
            while (l < r) {
                int mid = l + (r - l) / 2;
                if (tails[mid] < num) l = mid + 1;
                else r = mid;
            }
            tails[l] = num;
            if (result == r) result++;
        }

        return result;
    }

    public static void main(String[] args) {
        int[] nums = {4, 10, 4, 3, 8, 9};
        System.out.println(new Solution().lengthOfLIS(nums));
    }
}

File: 300_longest_increasing_subsequence/Solution2/Solution.py
class Solution:
    def lengthOfLIS(self, nums: list) -> int:
        arr_len = len(nums)
        if arr_len < 2: return arr_len

        tails = [0] * arr_len
        result = 0

        for num in nums:
            l, r = 0, result
            while l < r:
                m = (l + r) // 2
                if tails[m] < num: l = m + 1
                else: r = m
            tails[l] = num
            if result == r: result += 1

        return result


if __name__ == '__main__':
    nums = [4, 10, 4, 3, 8, 9]
    print(Solution().lengthOfLIS(nums))

